crossbar/worker/container.py

```diff
@@ -283,72 +326,20 @@ def panic(fail, msg):
             component = ContainerComponent(component_id, config, proto, None)
             self.components[component_id] = component
-
-            # FIXME: this is a total hack.
-            #
-            def close_wrapper(orig, was_clean, code, reason):
-                """
-                Wrap our protocol's onClose so we can tell when the component
-                exits.
-                """
-                r = orig(was_clean, code, reason)
-                if component.id not in self.components:
-                    self.log.warn("Component '{id}' closed, but not in set.",
-                                  id=component.id)
-                    return r
-
-                if was_clean:
-                    self.log.info("Closed connection to '{id}' with code '{code}'",
-                                  id=component.id, code=code)
-                else:
-                    self.log.error("Lost connection to component '{id}' with code '{code}'.",
-                                   id=component.id, code=code)
-
-                if reason:
-                    self.log.warn(str(reason))
-
-                del self.components[component.id]
-                self._publish_component_stop(component)
-                component._stopped.callback(component.marshal())
-
-                if not self.components:
-                    if self._exit_mode == self.SHUTDOWN_ON_LAST_COMPONENT_STOPPED:
-                        self.log.info("Container is hosting no more components: stopping container in exit mode <{exit_mode}> ...", exit_mode=self._exit_mode)
-                        self.shutdown()
-                    else:
-                        self.log.info("Container is hosting no more components: continue running in exit mode <{exit_mode}>", exit_mode=self._exit_mode)
-                else:
-                    self.log.info("Container is still hosting {component_count} components: continue running in exit mode <{exit_mode}>", exit_mode=self._exit_mode, component_count=len(self.components))
-
-                return r
-
-            # FIXME: due to history, the following is currently the case:
-            # ITransportHandler.onClose is implemented directly on WampWebSocketClientProtocol,
-            # while with WampRawSocketClientProtocol, the ITransportHandler is implemented
-            # by the object living on proto._session
-            #
+            # for RawSockets the onClose handler is attached to session.proto (above)
             if isinstance(proto, WampWebSocketClientProtocol):
                 proto.onClose = partial(close_wrapper, proto.onClose)
-
-            elif isinstance(proto, WampRawSocketClientProtocol):
-                # FIXME: doesn't work without guard, since proto_.session is not yet there when
-                # proto comes into existance ..
-                if proto._session:
-                    proto._session.onClose = partial(close_wrapper, proto._session.onClose)
-            else:
```

Comment:
Ok, my understanding of the problem was that the "if" was happening before proto._session had been instantiated, so that test in that position was obsolete as proto_session would be None, even if a session was in the process of being created .. so the assignment now happens "in" create_session, which only happens after the session has been created ... ?

Reply:
There should be one listener only, but it looks from the code like in websocket cases there will be TWO listeners attached: if it's a `WampWebSocketClientProtocol` it'll get the `close_wrapper` attached to `session.onClose` (because there's no "if" there) and also get `proto.onClose` attached (right?).

Maybe there are some CTS tests that cover this, I'll look -- I recall this stuff being really hard to get right but also can't remember the details, so want to make sure we don't break behavior :)